cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogAppender.java

```diff
@@ -77,33 +77,47 @@ private void addExtraTags(ILoggingEvent event) {
     String callerClass = callerData[0].getClassName();
 
     Map<String, String> tags = loggerExtraTags.getIfPresent(callerClass);
+    Class[] callerClasses = CallerClassSecurityManager.getCallerClasses();
     if (tags == null) {
-      tags = Collections.emptyMap();
-
-      for (Class<?> cls : CallerClassSecurityManager.getCallerClasses()) {
-        if (cls.getName().equals(callerClass)) {
-          String classLoaderName = cls.getClassLoader().getClass().getName();
-          switch (classLoaderName) {
-            case "co.cask.cdap.internal.app.runtime.plugin.PluginClassLoader":
-              tags = Collections.singletonMap(ORIGIN_KEY, "plugin");
-              break;
-            case "co.cask.cdap.internal.app.runtime.ProgramClassLoader":
-              tags = Collections.singletonMap(ORIGIN_KEY, "program");
-              break;
-            default:
-              tags = Collections.singletonMap(ORIGIN_KEY, "system");
-          }
-          break;
-        }
-      }
+      tags = addTagsForClass(callerClass, callerClasses);
 
       loggerExtraTags.put(callerClass, tags);
     }
 
+    // If no tags were added using callerData, try using LoggerName
+    if (tags.isEmpty()) {
+      tags = loggerExtraTags.getIfPresent(event.getLoggerName());
+      if (tags == null) {
+        tags = addTagsForClass(event.getLoggerName(), callerClasses);
+      }
+      loggerExtraTags.put(event.getLoggerName(), tags);
```

Comment:
yes, it saves us from trying over and over for some callerData/LoggerName that does not match with any frame on the execution stack. Although I should move the put inside the null check. Will update the diffs.